internal/allocator/allocator.go

```diff
@@ -269,31 +272,80 @@ func (a *Allocator) AllocateFromPool(svc string, serviceIPFamily ipfamily.Family
 }
 
 // Allocate assigns any available and assignable IP to service.
-func (a *Allocator) Allocate(svc string, serviceIPFamily ipfamily.Family, ports []Port, sharingKey, backendKey string) ([]net.IP, error) {
-	if alloc := a.allocated[svc]; alloc != nil {
-		if err := a.Assign(svc, alloc.ips, ports, sharingKey, backendKey); err != nil {
+func (a *Allocator) Allocate(svcKey string, svc *v1.Service, serviceIPFamily ipfamily.Family, ports []Port, sharingKey, backendKey string) ([]net.IP, error) {
+	pinnedPools := a.getSortedPinnedPools(svc)
+	for _, pool := range pinnedPools {
+		if ips, err := a.AllocateFromPool(svcKey, serviceIPFamily, pool.Name, ports, sharingKey, backendKey); err == nil {
+			return ips, nil
+		}
+	}
+	if alloc := a.allocated[svcKey]; alloc != nil {
+		if err := a.Assign(svcKey, alloc.ips, ports, sharingKey, backendKey); err != nil {
 			return nil, err
 		}
 		return alloc.ips, nil
 	}
 
-	for poolName := range a.pools {
-		if !a.pools[poolName].AutoAssign {
+	for poolName := range a.pools.ByName {
```

Comment:
can we iterate over pools (instead that over the keys), so we won't  need to dereference the map again?